Add a command to print the application's routing table

settermjd opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

This PR creates a new command that prints the application's routing table. For each route, it prints its name, path, middleware, and any additional options, in a tabular format, to the terminal. 

The motivation for the change was wanting to quickly look up this information while developing Mezzio-based applications, but having to trawl through configuration files to find it. It would be easier to have a command that collated and printed the information to the terminal, helping to save time and avoid missing any listed routes.

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

settermjd committed:
Add documentation for the new mezzio:routes:list command to the README

This change adds the command and it's core description to the README,
and also adds an extended documentation section to the end of the
README, showing what the command does, the options it supports, and some
sample output. The tabular output isn't quite correct, but is there to
at least have a placeholder.

settermjd committed:
Add the core classes for the command

This change adds the core classes, ListRoutesCommand and
ListRoutesCommandFactory, along with accompanying tests for them. At
this stage, the functionality is extremely limited, covering that the
command can be instantiated correctly, and has the correct
arguments/options, help output strings, command name, and description.
The rest of the functionality is coming.

settermjd committed:
Add a class that can filter a list of Route objects

This change adds a class that extends the FilterIterator from PHP's SPL.
The intention of the class is to be able to filter the routes in an
application's routing table by one or more of their four core
properties/attributes, those being name, path, method(s), and
middleware.

settermjd committed:
Add classes to sort a list of Routes by name and path

These two classes sort a traversable list of Route objects by both name
and path in ascending order.

settermjd committed:
Set declare strict in all classes

settermjd committed:
Fix code to meet the coding standard

boesing committed:
feature: add support for component-installer v3

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

boesing committed:
qa: import internal classes from component-installer into this package

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

boesing committed:
qa: import internal classes from component-installer into this package

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

weierophinney committed:
qa: adds tests for classes imported from laminas-component-installer

- `ConfigDiscovery\ConfigAggregatorTest`
- `ConfigInjector\AbstractInjectorTestCase`
- `ConfigInjector\ConfigAggregatorInjectorTest`
- related test assets

Additionally, the Register/Deregister command tests needed to mark several test methods to run in separate processes and ignore global state to ensure that class overriding would work.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

weierophinney committed:
qa: resolve errors flagged by psalm

- Marks internal classes and methods for suppression
- Identifies one `ArgumentTypeCoercion` in a test that is valid, and adds to the baseline.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

dependabot[bot] committed:
build(deps): Bump laminas/laminas-diactoros from 2.8.0 to 2.11.1

Bumps [laminas/laminas-diactoros](https://github.com/laminas/laminas-diactoros) from 2.8.0 to 2.11.1.
- [Release notes](https://github.com/laminas/laminas-diactoros/releases)
- [Commits](https://github.com/laminas/laminas-diactoros/compare/2.8.0...2.11.1)

---
updated-dependencies:
- dependency-name: laminas/laminas-diactoros
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add ability to render the routing table in tabular format

This change is quite simplistic, but renders the routing table in
tabular format if there are routes to render, otherwise it renders a
message saying that there are no routes available in the routing table.

settermjd committed:
Add the ability to format output as JSON

settermjd committed:
Only accept valid format options or no format option

This change updates the code to check that, if a format is supplied,
that it's one of the two available choices, table and json. If it's not
either of those, then an error message is printed.

settermjd committed:
Implement the ability to sort the routing table

The current options are by name and path. Other options may be added in
the future.

Ocramius commented:
12fc77a feels out of place - please rebase instead

settermjd committed:
Enable the ListRoutesCommand

This, small, change enables the ListRoutesCommand by adding it and the
command's name (mezzio:routes:list) in the ConfigProvider's
getConsoleConfig() function.

settermjd committed:
Enable the ListRoutesCommand

This, small, change enables the ListRoutesCommand by adding it and the
command's name (mezzio:routes:list) in the ConfigProvider's
getConsoleConfig() function.

settermjd committed:
Remove the option shortcuts

The intention here is to avoid having shortcuts that collide with
existing shortcuts, whether from other commands or from the core command
itself.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add the ListRoutesCommand in alphabetic order in the commands list

After I added the command (in a bit of a hurry) I then noticed that the
command were added in alphabetic order (best way to do it). So this
change fixes up my initial ordering mistake.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add the ListRoutesCommand in alphabetic order in the commands list

After I added the command (in a bit of a hurry) I then noticed that the
command were added in alphabetic order (best way to do it). So this
change fixes up my initial ordering mistake.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Replace the RouteCollector member variable with an array of Routes

I don't know why I didn't do this originally, but there was no real need
to have a RouteCollector variable (that I know of) as a member variable.
All the application needs is an array of Route objects, as that's all it
iterates over. This change corrects the original oversight.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Sort the routes before rendering them

Again, I don't know why I sorted the routes later, when they could have
been sorted earlier before rendering them. This change corrects it. I'll
put it down to not fully appreciating what I was trying to achieve at
first.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Refactor how RoutesFilter works

This change makes a number of small refactors:

- It doesn't attempt to filter if there are no filters to apply.
- It makes some fixes to the functionality for filtering routes by method.
- It adds functionality for filtering routes by middleware.
- It also renamed options to filterOptions with the intent of being more
  intuitive.

Around filtering by method, no further checks are made if any HTTP
method is allowed. This, I feel, improves performance by skipping
further checks.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Reformat the RouteSorterBy* classes

While not necessary, to me at least, reformatting them as this change
does makes them that much easier and quicker to read.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
> [12fc77a](https://github.com/mezzio/mezzio-tooling/commit/12fc77a99f63bc9f3abfdd87ffd9490752ec70d1) feels out of place - please rebase instead

Could I get an assist with that?

settermjd committed:
Finish implementing the routing filter options

This change finishes implementing the routing filter options, so that
the options passed from the command line when the command is called,
that those options will be respected. It also does a little bit of code
clean up so that the code is easier to read and maintain (I hope).

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

Ocramius commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 /coveralls-upload.json
 /phpunit.xml
 /vendor/
+.idea
```
Should be in your global gitignore: this artifact comes from your local tooling, not from the library.

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+use function strtolower;
+use function usort;
+
+class ListRoutesCommand extends Command
```
`final`

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+    private array $routes;
+
+    /**
+     * Document This!
```
Yup :P

settermjd replied:
Oops. That wasn't meant to slip in.

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+
+class ListRoutesCommand extends Command
+{
+    /** @var Route[]  */
```
Can we declare the key type too?

settermjd replied:
Will check and do that if I can.

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+    /**
+     * Document This!
+     *
+     * @var array
```
Let's declare the type of keys/values

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+        $this->setHelp(self::HELP);
+
+        $this->addOption(
+            'format',
```
These keys should be `private const` or such

settermjd replied:
I don't follow. Sorry.

froschdesign replied:
Examples:

```php
$this->addOption(
    self::OPTION_FORMAT,
    // …
);

$this->addOption(
    self::OPTION_SORT,
    // …
);

$this->addOption(
    self::OPTION_HAS_MIDDLEWARE,
    // …
);

$this->addOption(
    self::OPTION_HAS_NAME,
    // …
);
```

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+        usort($this->routes, $sorter);
+
+        $this->filterOptions = [
+            'method'     => strtolower((string) $input->getOption('supports-method')),
```
I suggest asserting on the values, rather than casting

boesing replied:
Yah, I would suggest using https://github.com/php-fig/http-message-util/blob/master/src/RequestMethodInterface.php and create proper assertions or return `ANY` which might be a constant only in this class

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+                    "Listing the application's routing table in table format."
+                );
+                break;
+            default:
+                $result = -1;
+                $output->writeln(
+                    "Invalid output format supplied. Valid options are 'table' and 'json'"
+                );
```
Perhaps just make it the default, or assert on `$format` before?

boesing replied:
Yah, I would prefer having `table` as a default. Thats basically how most of the composer CLI commands work as well 👍🏼 
Even options can contain defaults in `symfony/console` AFAIR. No need to have that empty string check.

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+    {
+        $rows = [];
+
+        /** @var Route[] $routesIterator */
```
This shouldn't be needed: something weird with type inference going on

Ocramius commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+        );
+
+        foreach ($routesIterator as $route) {
+            $routeMethods = implode(',', $route->getAllowedMethods() ?? []);
```
This `implode` looks fishy 🤔

boesing replied:
When changing this to something like this, its far more readable IMHO: 

```php
$routeMethods = $route->getAllowedMethods() ?? self::ANY_REQUEST_METHOD_COLLECTION;
```

Just add a constant like: 
```php
private const ANY_REQUEST_METHOD_COLLECTION = [
    RequestMethodInterface::METHOD_GET,
    // etc.
];
```

Then just use `implode` below. Afterwards, refactoring the code so we do not have `else` anymore might reduce complexity as well and then it might be understandable.

Ocramius commented on src/Routes/ListRoutesCommandFactory.php:
```diff
@@ -0,0 +1,18 @@
+{
+    public function __invoke(ContainerInterface $container): ListRoutesCommand
+    {
+        $routesCollector = $container->get(RouteCollector::class);
```
If you need the container to be typed, a stub may be added to this lib.

settermjd replied:
Not following, sorry.

boesing replied:
Psalm does not automagically understand that if you fetch something from the container what the return value is.
That's why psalm tells you that "mixed" is being assigned to the `$routesCollector` variable.

I usually avoid this by using this psalm plugin:

https://github.com/Lctrs/psalm-psr-container-plugin

Ocramius commented on src/Routes/Renderer/RendererInterface.php:
```diff
@@ -0,0 +1,10 @@
+
+interface RendererInterface
+{
+    public function render(array $routes): string;
```
The input format really could need some better specification

Ocramius commented on src/Routes/Sorter/RouteSorterByName.php:
```diff
@@ -0,0 +1,19 @@
+
+use Mezzio\Router\Route;
+
+class RouteSorterByName
```
In general, across the whole patch, please use `final` aggressively

Ocramius commented on src/Routes/Sorter/RouteSorterByName.php:
```diff
@@ -0,0 +1,19 @@
+            return 0;
+        }
+
+        return $routeOne->getName() < $routeTwo->getName() ? -1 : 1;
```
`$routeOne->getName() <=> $routeTwo->getName()` is the same as this method :)

Ocramius commented on src/Routes/Sorter/RouteSorterByPath.php:
```diff
@@ -0,0 +1,19 @@
+            return 0;
+        }
+
+        return $routeOne->getPath() < $routeTwo->getPath() ? -1 : 1;
```
`<=>` can be used here too

Ocramius commented on src/TemplateResolutionTrait.php:
```diff
@@ -88,6 +87,6 @@ private function getRendererServiceTypeFromContainer(ContainerInterface $contain
         }
         $renderer = $container->get(TemplateRendererInterface::class);
-        return get_class($renderer);
+        return $renderer::class;
```
Needs dropping PHP 7.4

settermjd replied:
Sorry, I made this change accidentally when running php-cs

Ocramius commented on test/Routes/ListRoutesCommandFactoryTest.php:
```diff
@@ -0,0 +1,41 @@
+
+class ListRoutesCommandFactoryTest extends TestCase
+{
+    use ProphecyTrait;
```
Let's use built-in mocks instead

settermjd replied:
Okay. Sure. Will do.

boesing replied:
Rule of thumb: do not use `prophecy` anymore. 😅

arueckauer commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,231 @@
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $result = 0;
+        if (empty($this->routes)) {
```
Since the `routes` property is an array, `count($this-routes) === 0` should be used instead to show clear intend. The use of `empty` is oftentimes confusing or misleading.

Ocramius replied:
Or `[] === $var` :+1:

settermjd committed:
Remove unrequired/unused classes

These classes were commented on during the PR which lead me to check and
remember that they're not used. Given that, this change removes them.

settermjd committed:
Remove .idea from .gitignore

This is being removed after feedback from @ocramius.

settermjd committed:
Implement feedback from PR review

boesing commented on src/Routes/RoutesFilter.php:
```diff
@@ -0,0 +1,170 @@
+ * RoutesFilter filters a traversable list of Route objects based on any of the four Route criteria,
+ * those being the route's name, path, middleware, or supported method(s).
+ */
+class RoutesFilter extends FilterIterator
```
you should provide what the iterator is actually about. Usually, this is done via generics:

```php
/**
 * @template-extends FilterIterator<array-key,Route>
 */
final class RoutesFilter extends FilterIterator
{}
```

I just used `array-key` because I guess we don't mind about the key type.

boesing commented on src/Routes/RoutesFilter.php:
```diff
@@ -0,0 +1,170 @@
+     */
+    private array $filterOptions;
+
+    public function __construct(ArrayIterator $routes, array $filterOptions = [])
```
You should pass what type the `ArrayIterator` is of. In this case, it might be: 

```php
/**
 * @param ArrayIterator<array-key,Route>
 */
```

boesing replied:
What about the `filterOptions`. 
Can we probably move that stuff into a dedicated object so we do actually know what "options" are available?

boesing commented on src/Routes/RoutesFilter.php:
```diff
@@ -0,0 +1,170 @@
+
+        $this->filterOptions = $filterOptions;
+
+        // Filter out any options that are, effectively, "empty".
```
You would not need to filter anything if you provide a proper option object. Can we probably change the code so we get rid of that array?

settermjd committed:
Revert an unintentional change to TemplateResolutionTrait

The change that this commit reverts was unintentionally made when
running php-cbf. This change restores the original functionality as PHP
7.4 is still supported by the library.Revert an unintentional change to
  TemplateResolutionTrait

  The change that this commit reverts was unintentionally made when
  running php-cbf. This change restores the original functionality as
  PHP 7.4 is still supported by the library.

renovate[bot] committed:
Add renovate.json

Commit:
remove deprecated psalm configuration

Signed-off-by: Gary Lockett <gary@creativecow.uk>

renovate[bot] committed:
Update dependency psalm/plugin-phpunit to ^0.17.0

| datasource | package              | from   | to     |
| ---------- | -------------------- | ------ | ------ |
| packagist  | psalm/plugin-phpunit | 0.16.1 | 0.17.0 |


Signed-off-by: Renovate Bot <bot@renovateapp.com>

settermjd commented:
I'm not sure how to finish the PR because of how the routes are typically loaded in an application generated with the [Mezzio Skeleton](https://github.com/mezzio/mezzio-skeleton). What I mean is that the routing table is typically loaded in _public/index.php_, using the following code:

```php
(function () {
    /** @var \Psr\Container\ContainerInterface $container */
    $container = require 'config/container.php';

    /** @var \Mezzio\Application $app */
    $app = $container->get(\Mezzio\Application::class);
    $factory = $container->get(\Mezzio\MiddlewareFactory::class);

    // Execute programmatic/declarative middleware pipeline and routing
    // configuration statements
    (require 'config/pipeline.php')($app, $factory, $container);
    (require 'config/routes.php')($app, $factory, $container);

    $app->run();
})();
```

So, how should the code load the routes so that it can get access to the information?

weierophinney commented:
> So, how should the code load the routes so that it can get access to the information?

Since commands are typically run in the project root, you can generally assume that you can load the routes file using `require 'config/routes.php';`. This file, and the `config/pipeline.php` file, returns an anonymous function, which accepts three arguments:

- A `Mezzio\Application` instance
- A `Mezzio\MiddlewareFactory` instance
- The PSR-11 container instance

Once that function has been called, the `Mezzio\Router\RouteCollector` instance should be fully populated, and you can loop over its `getRoutes()` method to introspect the various `Mezzio\Router\Route` instances.

Your command will need to either accept all three of the above objects to its constructor, or you could have it accept just the container, and then pull the other two from the container when you are ready to create the routing table.

As a minimal example:

```php
class RoutingTableCommand extends Commands
{
    public function __construct(private ContainerInterface $container)
    {
        parent::__construct();
    }

    protected function configure(): void
    {
         // setup command information here, such as options or arguments
    }

    protected function execute(InputInterface $input, OutputInterface $outptu): int
    {
        // create routing table
        (require 'config/routes.php')(
            $this->container->get('Mezzio\Application'),
            $this->container->get('Mezzio\MiddlewareFactory'),
            $this->container
        );

        $routes = $container->get('Mezzio\Router\RouteCollector');

        // iterate over the routes
        foreach ($routes->getRoutes() as $route) {
        }

        return 0;
    }
}
```

boesing commented:
Can we make this configurable? We do have a routes.php in combination with the https://github.com/mezzio/mezzio/blob/3.12.x/src/Container/ApplicationConfigInjectionDelegator.php 
If you now expect all routes.php to return a callables, that would be too specific. At least check if the config has the delegator registered and if so, do not require the routes.php.

edit: could be a routeloaderinterface which is either a "nooploader" or a "callables routes.php loader" for example.

settermjd commented:
> Can we make this configurable? We do have a routes.php in combination with the https://github.com/mezzio/mezzio/blob/3.12.x/src/Container/ApplicationConfigInjectionDelegator.php If you now expect all routes.php to return a callables, that would be too specific. At least check if the config has the delegator registered and if so, do not require the routes.php.
> 
> edit: could be a routeloaderinterface which is either a "nooploader" or a "callables routes.php loader" for example.

Mind if we have a call to talk about this further?

boesing commented:
Sure! I have to travel to munich this evening. Maybe we find some time later or on Sunday/next week?

boesing commented:
Since we did not made a call yet, I took some time to looked into the actual PR.

IMHO, the command itself is properly implemented. Thats actually the exact same way as I would've done that. The commands factory receives the route collector. So thats already nice.

When I see this correct, we need the following tasks finished so that we can merge this:

- [ ] getting rid of the psalm issues, some of them can be fixed by adding [this psalm plugin](https://github.com/Lctrs/psalm-psr-container-plugin)
- [x] replacement of prophecy with mocks
- [x] proper namespaces (unit tests are failing as the namespace of some classes is incorrect)

Almost every check is red here, so maybe getting them green would be fine.

---


Another thing is - I already provided a command in laminas-cache. This command is kinda related to `mezzio-router`, so is there a reason why we do not put it in that library?

There are other commands in this library which do not properly fit into a specific laminas/mezzio component as they do not provide anything for a specific component (`CreateHandler`/`CreateMiddleware` do generate code for PSR-15, `Migrate*` commands are related to PSR-15 as well).

The Module stuff could be part of  `mezzio` itself since we migrated this to `laminas-cli`.

Just curious - no need to change this right now, but I would prefer having stuff in the components they belong to. That would prevent having to install mezzio-tooling just for the route table command.

boesing commented:
Oh, there comes to my mind, that there are multiple ways to configure the app to have routes...

This is another way:

```php
(function () {
    /** @var \Psr\Container\ContainerInterface $container */
    $container = require 'config/container.php';

    /** @var \Mezzio\Application $app */
    $app = $container->get(\Mezzio\Application::class);
    $factory = $container->get(\Mezzio\MiddlewareFactory::class);

    // Execute programmatic/declarative middleware pipeline and routing
    // configuration statements
    (require 'config/pipeline.php')($app, $factory, $container);
    (require 'config/routes.php')($app, $factory, $container);

    $app->get('/i-am-route', [IAmRouteRequestHandler::class], 'i-am-route');

    $app->run();
})();
```

So how should we receive that route which is being added via `$app->get()`? Imho, the way how mezzio is "configurable" in that way, its impossible to provide a proper command for all kind of project configurations.

So to summarize, there are:

1. applications using `routes.php` which returns an array (afair that was the way how expressive was configured in either its 1st version or in even older versions) and/or `ConfigProvider` (thats actually how we do it in our company, so we have both `routes.php` and some routes are part of `ConfigProvider`)
2. applications using `routes.php` which returns a callable (consuming application, container, etc.)
3. applications using `ConfigProvider` along with delegators for `Application`


IMHO, thats quite a lot of ways. Does anyone have an idea on how to be compatible with all these types of configurations?

gsteel commented:
FWIW, I normally setup routes by delegating around `RouteCollector` so that on the cli, if I need to generate routes, I can just grab the router from the container in the knowledge that all my routes are configured without needing to bootstrap the application - typically, I delete `routes.php`. In the past I've also used delegators on `Application` to inject routes

boesing commented:
@gsteel thats exactly the point, there are multiple ways of providing routes.

Since it is impossible to actually catch index.php routes (be it callables router.php or direct Application#get), I would try to Focus on just loading the route collector.

I like the idea of this command, but I wouldnt pay attention to the bunch of ways to provide a route.

Adding some description to the command might be good enough. Routes can be still passed to a delegator if some1 wants to use the command.

renovate[bot] committed:
Update all non-major dependencies

| datasource | package              | from   | to     |
| ---------- | -------------------- | ------ | ------ |
| packagist  | psalm/plugin-mockery | 0.9.1  | 0.10.0 |
| packagist  | psalm/plugin-phpunit | 0.17.0 | 0.18.0 |


Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Lock file maintenance

Signed-off-by: Renovate Bot <bot@renovateapp.com>

renovate[bot] committed:
Lock file maintenance

Signed-off-by: Renovate Bot <bot@renovateapp.com>

ghostwriter committed:
Update composer config `platform.php` to `8.0.99`

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

ghostwriter committed:
Ignore PHP platform requirements via `.laminas-ci.json`

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

ghostwriter committed:
Update composer `php` to `~8.0.0 || ~8.1.0 || ~8.2.0`

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

ghostwriter committed:
Update code to PHP 8

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

ghostwriter committed:
Update psalm-baseline.xml

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

ghostwriter committed:
Fix CS and Psalm issues

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

renovate[bot] committed:
Update dependency psalm/plugin-mockery to ^0.11.0

| datasource | package              | from   | to     |
| ---------- | -------------------- | ------ | ------ |
| packagist  | psalm/plugin-mockery | 0.10.0 | 0.11.0 |


Signed-off-by: Renovate Bot <bot@renovateapp.com>

Ocramius committed:
Rewrote `RegisterCommand` and its tests to no longer depend on global state

Ocramius committed:
Rewrote `DeregisterCommand` and its tests to no longer depend on global state

Ocramius committed:
Dropping all outdated dependencies, upgrading to latest of all

This still respects dependencies that require PHP 8.0 to run, due to `config.platform.php`
being set.

Ocramius committed:
Rewrote test suite to get rid of Prophecy

Ocramius committed:
Updated psalm baseline to reflect test suite changes after prophecy removal

gsteel committed:
Add newly discovered psalm issues to the baseline

Signed-off-by: George Steel <george@net-glue.co.uk>

renovate[bot] committed:
Update dependency laminas/laminas-coding-standard to ~2.5.0

| datasource | package                         | from  | to    |
| ---------- | ------------------------------- | ----- | ----- |
| packagist  | laminas/laminas-coding-standard | 2.4.0 | 2.5.0 |


Signed-off-by: Renovate Bot <bot@renovateapp.com>

settermjd commented:
Sorry about dropping this. Back on it now.

settermjd commented:
> > So, how should the code load the routes so that it can get access to the information?
> 
> Since commands are typically run in the project root, you can generally assume that you can load the routes file using `require 'config/routes.php';`. This file, and the `config/pipeline.php` file, returns an anonymous function, which accepts three arguments:
> 
>     * A `Mezzio\Application` instance
> 
>     * A `Mezzio\MiddlewareFactory` instance
> 
>     * The PSR-11 container instance
> 
> 
> Once that function has been called, the `Mezzio\Router\RouteCollector` instance should be fully populated, and you can loop over its `getRoutes()` method to introspect the various `Mezzio\Router\Route` instances.
> 
> Your command will need to either accept all three of the above objects to its constructor, or you could have it accept just the container, and then pull the other two from the container when you are ready to create the routing table.
> 
> As a minimal example:
> 
> ```
> class RoutingTableCommand extends Commands
> {
>     public function __construct(private ContainerInterface $container)
>     {
>         parent::__construct();
>     }
> 
>     protected function configure(): void
>     {
>          // setup command information here, such as options or arguments
>     }
> 
>     protected function execute(InputInterface $input, OutputInterface $outptu): int
>     {
>         // create routing table
>         (require 'config/routes.php')(
>             $this->container->get('Mezzio\Application'),
>             $this->container->get('Mezzio\MiddlewareFactory'),
>             $this->container
>         );
> 
>         $routes = $container->get('Mezzio\Router\RouteCollector');
> 
>         // iterate over the routes
>         foreach ($routes->getRoutes() as $route) {
>         }
> 
>         return 0;
>     }
> }
> ```

Thanks for this, @weierophinney. Getting back in to this to bring it to a close.

settermjd commented:
> @gsteel thats exactly the point, there are multiple ways of providing routes.
> 
> Since it is impossible to actually catch index.php routes (be it callables router.php or direct Application#get), I would try to Focus on just loading the route collector.
> 
> I like the idea of this command, but I wouldnt pay attention to the bunch of ways to provide a route.
> 
> Adding some description to the command might be good enough. Routes can be still passed to a delegator if some1 wants to use the command.

So, it seems that the command needs to be configurable by the user, so that they tell it where to look?

settermjd committed:
Update files to match the coding standard

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

boesing commented:
@settermjd no I would simply load the route collector and extract all known methods from there. if a project adds routes via index.php, these projects cant use this command until they move routes to configuration or to delegators. But let that be the problem of those projects. Id say focus on runtime extraction.

settermjd committed:
Remove reference to ProphecyTrait

As the project doesn't use it anymore, it's time to remove it.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Further refactor of code from ProphecyTrait

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
> > So, how should the code load the routes so that it can get access to the information?
> 
> Since commands are typically run in the project root, you can generally assume that you can load the routes file using `require 'config/routes.php';`. This file, and the `config/pipeline.php` file, returns an anonymous function, which accepts three arguments:
> 
>     * A `Mezzio\Application` instance
> 
>     * A `Mezzio\MiddlewareFactory` instance
> 
>     * The PSR-11 container instance
> 
> 
> Once that function has been called, the `Mezzio\Router\RouteCollector` instance should be fully populated, and you can loop over its `getRoutes()` method to introspect the various `Mezzio\Router\Route` instances.
> 
> Your command will need to either accept all three of the above objects to its constructor, or you could have it accept just the container, and then pull the other two from the container when you are ready to create the routing table.
> 
> As a minimal example:
> 
> ```
> class RoutingTableCommand extends Commands
> {
>     public function __construct(private ContainerInterface $container)
>     {
>         parent::__construct();
>     }
> 
>     protected function configure(): void
>     {
>          // setup command information here, such as options or arguments
>     }
> 
>     protected function execute(InputInterface $input, OutputInterface $outptu): int
>     {
>         // create routing table
>         (require 'config/routes.php')(
>             $this->container->get('Mezzio\Application'),
>             $this->container->get('Mezzio\MiddlewareFactory'),
>             $this->container
>         );
> 
>         $routes = $container->get('Mezzio\Router\RouteCollector');
> 
>         // iterate over the routes
>         foreach ($routes->getRoutes() as $route) {
>         }
> 
>         return 0;
>     }
> }
> ```

I know it's been 2 years since I last got on this, but that worked a treat, @mwop.

settermjd committed:
Introduced a class to load routes from config/routes.php

After some discussion and thought about how to do this, @ghostwriter
suggested extracting the logic of loading routes from config/routes.php
into an external class so that it's easy to test. This change introduces
the class that implements a simple interface to do that.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Refactor how routes are loaded

This change makes use of the new RoutesFileConfigLoader to load any
routes defined in config/routes.php and then loads any routes defined in
any ConfigProvider classes. This was advice recommended by @mwop in this
comment
https://github.com/mezzio/mezzio-tooling/pull/27#issuecomment-1258541486

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Format files to pass phpcs rules

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

ghostwriter commented on README.md:
```diff
@@ -37,12 +37,59 @@ This package exposes commands for [laminas-cli](https://docs.laminas.dev/laminas
+| `--sort`            | `-s`          | Sort the command's output. The supported values are `name` and `path`.                                                                                                                                                                                                                                                           |
+| `--supports-method` | `-m`          | Accepts a comma-separated list of one or more HTTP methods, and filters out routes which don't support those methods.                                                                                                                                                                                                            |
+| `--has-path`        | `-p`          | Accepts a comma-separated list of one or more paths, and filters out routes with paths that don't match. The paths can be a regular expression, supported by the preg_* functions. For example, "/,/api/ping,*/ping".                                                                                                            |
+| `--has-name`        | `-n`          | Accepts a comma-separated list of one or more names, and filters out routes with names that don't match. The names can be fixed strings, or regular expressions supported by the preg_* functions. For example, "user,user.register,*.register,user*".                                                                           |
```
```suggestion
| `--has-name`        | `-n`          | Accepts a comma-separated list of one or more names, and filters out routes with names that don't match. The names can be fixed strings, or regular expressions supported by the "preg_*" functions. For example, "user,user.register,*.register,user*".                                                                           |
```

Fixes `Error: README.md:63:207 MD037/no-space-in-emphasis Spaces inside emphasis markers [Context: "* f”]`

ghostwriter commented on README.md:
```diff
@@ -37,12 +37,59 @@ This package exposes commands for [laminas-cli](https://docs.laminas.dev/laminas
+| `--supports-method` | `-m`          | Accepts a comma-separated list of one or more HTTP methods, and filters out routes which don't support those methods.                                                                                                                                                                                                            |
+| `--has-path`        | `-p`          | Accepts a comma-separated list of one or more paths, and filters out routes with paths that don't match. The paths can be a regular expression, supported by the preg_* functions. For example, "/,/api/ping,*/ping".                                                                                                            |
+| `--has-name`        | `-n`          | Accepts a comma-separated list of one or more names, and filters out routes with names that don't match. The names can be fixed strings, or regular expressions supported by the preg_* functions. For example, "user,user.register,*.register,user*".                                                                           |
+| `--has-middleware`  | `-w`          | Accepts a comma-separated list of one or more middleware classes, and filters out routes that do not require those classes. The classes can be fully-qualified, unqualified, or a regular expression, supported by the preg_* functions. For example, "\Mezzio\Middleware\LazyLoadingMiddleware,LazyLoadingMiddleware,\Mezzio*". |
```
```suggestion
| `--has-middleware`  | `-w`          | Accepts a comma-separated list of one or more middleware classes, and filters out routes that do not require those classes. The classes can be fully-qualified, unqualified, or a regular expression, supported by the "preg_*" functions. For example, "\Mezzio\Middleware\LazyLoadingMiddleware,LazyLoadingMiddleware,\Mezzio*". |
```

Fixes: `Error: README.md:65:261 MD037/no-space-in-emphasis Spaces inside emphasis markers [Context: "* f”]`

ghostwriter reviewed (commented):
To update or regenerate the psalm baseline file.

`vendor/bin/psalm --no-diff --update-baseline=psalm-baseline.xml`
or
`vendor/bin/psalm --no-diff --set-baseline=psalm-baseline.xml`

settermjd committed:
feature: add support for component-installer v3

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

settermjd committed:
Disable the rendering of emojis in Markdown

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

settermjd committed:
add phpdocumentor/reflection-docblock >=4.0.1 requirement

Signed-off-by: Gary Lockett <gary@creativecow.uk>

settermjd committed:
Update dependency laminas/laminas-coding-standard to ~2.4.0

| datasource | package                         | from  | to    |
| ---------- | ------------------------------- | ----- | ----- |
| packagist  | laminas/laminas-coding-standard | 2.3.0 | 2.4.0 |

Signed-off-by: Renovate Bot <bot@renovateapp.com>

settermjd committed:
Update all non-major dependencies

| datasource | package                   | from   | to     |
| ---------- | ------------------------- | ------ | ------ |
| packagist  | mikey179/vfsstream        | 1.6.11 | 1.6.11 |
| packagist  | mockery/mockery           | 1.5.1  | 1.5.1  |
| packagist  | php-mock/php-mock-phpunit | 2.6.1  | 2.6.1  |
| packagist  | phpspec/prophecy-phpunit  | 2.0.1  | 2.0.1  |
| packagist  | phpunit/phpunit           | 9.5.25 | 9.5.26 |
| packagist  | vimeo/psalm               | 4.29.0 | 4.29.0 |

Signed-off-by: Renovate Bot <bot@renovateapp.com>

settermjd committed:
Update composer.lock

Signed-off-by: Nathanael Esayeas <nathanael.esayeas@protonmail.com>

settermjd committed:
Dropped `phpspec/prophecy` from dependencies

settermjd committed:
Update dependency vimeo/psalm to v5

| datasource | package     | from   | to    |
| ---------- | ----------- | ------ | ----- |
| packagist  | vimeo/psalm | 4.30.0 | 5.0.0 |

Signed-off-by: Renovate Bot <bot@renovateapp.com>

settermjd committed:
Minor updates for markdownlint

settermjd committed:
Reset Psalm baseline

Doing this as per suggestion from @ghostwriter
(https://github.com/mezzio/mezzio-tooling/pull/27#pullrequestreview-2189976753).

settermjd committed:
Refactor matchesByRegex to return only a bool

After running Psalm, it became clear that the potential types that the
function could return were too many and unnecessary. This change
simplifies them down to just one, a boolean, which is much clearer.

settermjd committed:
Revise docblocks to pass Psalm static analysis

settermjd commented:
This PR has (to me) become a bit of a mess. What is the best way to finish it? Should I close it and start over on a newer branch?

alexmerlin commented:
IMO a new PR on a different branch - this PR/branch has been through a lot over the past couple of years.

Btw, we also have this feature in Dotkernl API, pretty much the same logic, slightly different implementation.

settermjd commented:
> IMO a new PR on a different branch - this PR/branch has been through a lot over the past couple of years.
> 
> Btw, we also have this feature in Dotkernl API, pretty much the same logic, slightly different implementation.

Thanks @alexmerlin. I'm looking to close this PR and create a new one (hopefully getting it merged quickly).

settermjd commented:
Closing in favour of a newer, cleaner PR.
